rfcs/0000-slsa-attestation.md

```diff
@@ -0,0 +1,365 @@
+          "content": "[{\"name\":\"my-registry-creds\",\"resourceVersion\":\"53\"},{\"name\":\"my-git-creds\"}]"
+        },
+      ]
+    },
+    "metadata": {
+      "invocationID": "https://kpack.io/default/test-build-1/test-build-1-build-pod@gke-test-default-pool-0582cba3-2dp4",
+      "startedOn": "2023-11-09T16:47:17.215551-05:00",
+      "finishedOn": "2023-11-09T16:54:17.215551-05:00"
+    },
+    "byproducts": []
+  }
+}
+```
+
+#### Statement
+The provenance is wrapped in an [in-toto statement][in-toto-statement] and formated as JSON. The `subject.name` and
+`subject.digest` should be set to the built image and digest.
+```json
+{
+  "_type": "https://in-toto.io/Statement/v0.1",
+  "predicateType": "https://slsa.dev/provenance/v1",
+  "predicate": <provenance from previous step>
+  "subject": [
+    {
+      "name": "gcr.io/cf-build-service-public/kpack/controller",
+      "digest": {
+        "sha256": "sha256:5e743affae0df727dce82df3f4fe5256cd66f41d919b3cd2aa91605eea2371f8"
+      }
+    }
+  ],
+}
+```
+#### Envelope
+
+The statement is stored in a [DSSE envelope][dsse-envelope], the signatures field may be empty.
+```json
+{
+  "payloadType": "application/vnd.in-toto+json",
+  "payload": <base64 encoded statement base previous step>
+  "signatures": [
+    {
+      "keyid": "default/my-cosign-secret",
+      "sig": "MEUCIQD/aWTUPVTRhWoGv1jYAvrnmYRJQmHVdy4NrmmLIxUaaAIgSumlFxSX9FG/wfbpYUAQJtE1/vzfVtRXmlr2LwpU670="
+    }
+  ]
+}
+```
+
+#### Storage
+
+The attestation is attached to the target image tag (as well as any `additionalTags`) using the
+[cosign attestation spec][cosign-attestation], where the envelope becomes a single layer. The attestation image also
+follows the [cosign tag-based discovery][cosign-tag] except that the suffix is `.att` instead of `.sig`.
+
+For example,
+`gcr.io/cf-build-service-public/kpack/controller@sha256:5e743affae0df727dce82df3f4fe5256cd66f41d919b3cd2aa91605eea2371f8`
+will have its corresponding attestation saved to
+`gcr.io/cf-build-service-public/kpack/controller:sha256-5e743affae0df727dce82df3f4fe5256cd66f41d919b3cd2aa91605eea2371f8.att`
+
+[slsa-recommended-suite]: https://slsa.dev/attestation-model#recommended-suite
+[slsa-schema]: https://slsa.dev/spec/v1.0/provenance#schema
+[in-toto-statement]: https://github.com/in-toto/attestation/blob/main/spec/v1/statement.md
+[dsse-envelope]: https://github.com/secure-systems-lab/dsse/blob/v1.0.0/envelope.md
+[cosign-attestation]: https://github.com/sigstore/cosign/blob/main/specs/ATTESTATION_SPEC.md
+[cosign-tag]: https://github.com/sigstore/cosign/blob/main/specs/SIGNATURE_SPEC.md#tag-based-discovery
+
+### Key types
+
+At minimum, the implementation should support the following key types:
+- [cosign key][cosign-key-format]. This is to ensure interop with `cosign verify-attestation`.
+- ECDSA private keys. This is due to it being the recommended SLSA suite.
+- RSA private keys. This is due to it being an extreamly popular key type.
```

Comment:
See the additional considerations down below, tldr: the only verification tool that works with key-based attestation is `cosign`, but they don't support non cosign-format keys. So there's no (nice) way of verifying ECDSA and RSA key signatures.